lib/Service/MailManager.php

```diff
@@ -509,7 +511,50 @@ public function deleteMailbox(Account $account,
 	 * @param Message $message
 	 * @return array[]
 	 */
-	public function getMailAttachments(Account $account, Mailbox $mailbox, Message $message) : array {
+	public function getMailAttachments(Account $account, Mailbox $mailbox, Message $message): array {
 		return $this->imapMessageMapper->getAttachments($this->imapClientFactory->getClient($account), $mailbox->getName(), $message->getUid());
 	}
+
+	/**
+	 * Filter out IMAP flags that aren't supported by the client server
+	 *
+	 * @param Horde_Imap_Client_Socket $client
+	 * @param string $flag
+	 * @param string $mailbox
+	 * @return array
+	 */
+	public function filterFlags(Account $account, string $flag, string $mailbox): array {
+		// check for RFC server flags
+		if (array_key_exists($flag, self::ALLOWED_FLAGS) === true) {
+			return self::ALLOWED_FLAGS[$flag];
+		}
+
+		// Only allow flag setting if IMAP supports Permaflags
+		// @TODO check if there are length & char limits on permflags
+		if ($this->isPermflagsEnabled($account, $mailbox) === true) {
+			return ["$" . $flag];
+		}
+		return [];
+	}
+
+	/**
+	 * Check IMAP server for support for PERMANENTFLAGS
+	 *
+	 * @param Account $account
+	 * @param string $mailbox
+	 * @return boolean
+	 */
+	public function isPermflagsEnabled(Account $account, string $mailbox): bool {
+		$client = $this->imapClientFactory->getClient($account);
+		try {
+			$capabilities = $client->status($mailbox, Horde_Imap_Client::STATUS_PERMFLAGS);
+		} catch (Horde_Imap_Client_Exception $e) {
+			throw new ServiceException(
+				"Could not get message flag options from IMAP: " . $e->getMessage(),
+				(int) $e->getCode(),
+				$e
+			);
+		}
+		return ($capabilities !== null && array_key_exists('permflags', $capabilities) === true && in_array("\*", $capabilities['permflags']) === true);
```

Comment:
```suggestion
		return ($capabilities !== null && array_key_exists('permflags', $capabilities) === true && in_array("\*", $capabilities['permflags'], true) === true);
```
please check if a strict comparison is possible: https://www.php.net/manual/en/function.in-array.php